src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java

```diff
@@ -123,20 +132,51 @@ static String generateAppInstallationToken(String appId, String appPrivateKey, S
                 appInstallation = appInstallations.get(0);
             } else {
                 appInstallation = appInstallations.stream()
-                        .filter(installation -> installation.getAccount().getLogin().equals(owner))
-                        .findAny()
-                        .orElseThrow(() -> new IllegalArgumentException(String.format(ERROR_NOT_INSTALLED, appId)));
+                    .filter(installation -> installation.getAccount().getLogin().equals(owner))
+                    .findAny()
+                    .orElseThrow(() -> new IllegalArgumentException(String.format(
+                        ERROR_NOT_INSTALLED,
+                        appId)));
             }
 
             GHAppInstallationToken appInstallationToken = appInstallation
-                    .createToken(appInstallation.getPermissions())
-                    .create();
-
-            return appInstallationToken.getToken();
-        } catch (IOException e) {
+                .createToken(appInstallation.getPermissions())
+                .create();
+
+            long expiration = getExpirationSeconds(appInstallationToken);
+            LOGGER.log(Level.FINEST,
+                "Token raw expiration epoch seconds: {0,number,#}",
+                expiration);
+
+            AppInstallationToken token = new AppInstallationToken(appInstallationToken.getToken(),
+                expiration);
+            LOGGER.log(Level.FINE,
+                "Generated App Installation Token for app ID {0}",
+                appId);
+
+            return token;
+        } catch (Exception e) {
+            LOGGER.log(Level.WARNING, "Failed to retrieve GitHub App installation token for app ID " + appId, e);
             throw new IllegalArgumentException(String.format(ERROR_AUTHENTICATING_GITHUB_APP, appId), e);
         }
+    }
+
+    private static long getExpirationSeconds(GHAppInstallationToken appInstallationToken) {
+        try {
+            return appInstallationToken.getExpiresAt()
+                .toInstant()
+                .getEpochSecond();
+        } catch (Exception e) {
```

Comment:
What exception is expected here?

Reply:
I was confused about this earlier too. `IOException` is declared to be thrown by [GHAppInstallationToken.getExpiresAt](https://github.com/hub4j/github-api/blob/fc38dba59a9608e1aaa182c99090a1fe1ab53329/src/main/java/org/kohsuke/github/GHAppInstallationToken.java#L139), but looking at [GitHubClient.parseDate](https://github.com/hub4j/github-api/blob/fc38dba59a9608e1aaa182c99090a1fe1ab53329/src/main/java/org/kohsuke/github/GitHubClient.java#L651) I think the `throws` clause on `GHAppInstallationToken.getExpiresAt` could just be dropped. Date parsing could throw runtime exceptions like `DateTimeParseException` (not sure what would cause it in practice, perhaps changes on GitHub's side), so maybe that is the kind of thing this is trying to catch.

Reply:
Implementation has changed over time, but changing the `throws` is an API change that will cause compilation to fail, right?

It is also possible for GitHub to hand us values that don't parse.  Regardless of why it fails, we want to recover and continue.